upstairs/src/guest.rs

```diff
@@ -413,13 +413,25 @@ impl Guest {
         Ok(())
     }
 
-    async fn backpressure_sleep(&self) {
-        let bp =
-            Duration::from_micros(self.backpressure_us.load(Ordering::SeqCst));
-        if bp > Duration::ZERO {
-            let _guard = self.backpressure_lock.lock().await;
-            tokio::time::sleep(bp).await;
-            drop(_guard);
+    /// Sleeps for a backpressure-dependent amount, holding the lock
+    ///
+    /// If backpressure is saturated, logs and returns an error.
+    async fn backpressure_sleep(&self) -> Result<(), CrucibleError> {
+        let bp = self.backpressure.load();
+        match bp {
+            BackpressureAmount::Saturated => {
+                let err = "write queue is saturated";
+                error!(self.log, "{err}");
```

Comment:
I wonder a few things here.

1) If we are saturated, what impact will we see from logging a bunch of messages.  Like, will it further impact our ability to process IOs?  Or, are things so bad at this point that it does not matter that we are spending time logging things?

2) I do like the idea of knowing this has happened, and if we don't log a message then bumping some internal counter I think would be good as it could help us explain why some VM got a bunch of errors yesterday but now things look okay.

3) And, finally, how difficult is it to trigger this condition from a VM?

Reply:
1. In the saturated case, perhaps we should still sleep for `delay_max` here before returning the error code?  That would limit us to one error every 30 seconds, which seems perfectly reasonable.  We could also log an error then delay some long amount of time (e.g. 1 hour), which would match our previous behavior but seems fraught; I worry that delaying a guest by 1 hour would be more confusing than just returning an error.

2. Agreed, we should definitely record it somewhere if not in an error message

3. From a VM, you should only see this error condition if writes take > 30 seconds to execute in Crucible.  The VM can't really affect this (assuming it's already spamming IOs as quickly as possible); the saturated error indicates that even at maximum backpressure, IOs are not being answered fast enough.